nhs/data/join_on_sa1.py

```diff
@@ -0,0 +1,219 @@
+            c=area_result_df["attribute"].to_numpy(),
+            cmap="tab10",
+            s=50,
+            alpha=0.7,
+        )
+        plt.title(
+            f"Final Positions after Random Distribution for Area Code: {area_code}"
+        )
+        plt.xlabel("Final Longitude")
+        plt.ylabel("Final Latitude")
+
+        plt.tight_layout()
+        plt.show()
+
+
+def main() -> None:
+    config_path = "configurations.yaml"
+    config = read_config(config_path)
+
+    shapefile_path = config["shapefile_path"]
+    csv_path = config["csv_path"]
+    crs = config["crs"]
+
+    total_rows = read_shapefile(shapefile_path)
+    houses_df = read_csv(csv_path)
+
+    pnts_gdf = prepare_points(houses_df, crs)
+    map_data = gpd.read_file(shapefile_path).to_crs(pnts_gdf.crs)
+
+    num_cores = config.get("num_cores", 4)  # Default to 4 if not specified
+    final_result = parallel_process(pnts_gdf, map_data, num_cores)
+
+    if (
+        final_result["area"].is_null().any()
+        or final_result["area_code"].is_null().any()
+    ):
+        logger.warning(
+            "Some points were not assigned to an SA1 area even after nearest join. Please check your data."
+        )
+
+    houses_with_areas = houses_df.join(final_result, on=["x", "y"], how="left")
+
+    output_csv_path = config["output_csv_path"]
+    houses_with_areas.write_csv(output_csv_path)
+    logger.info(f"CSV file saved to: {output_csv_path}")
+
+    missing_area_count = final_result["area"].is_null().sum()
+    logger.info(f"Number of missing values in the 'area' column: {missing_area_count}")
+
+    houses_with_attributes = houses_with_areas.with_columns(
+        pl.Series("attribute", np.random.randint(1, 11, size=len(houses_with_areas)))
+    )
+
+    # Check if random distribution is enabled in the config
+    if config.get("enable_random_distribution", False):
+        area_codes = houses_with_attributes["area_code"].unique().to_list()[:4]
+        result_dfs = []
+
+        for area_code in area_codes:
+            logger.info(f"Running random distribution for area code: {area_code}")
+            area_df = houses_with_attributes.filter(pl.col("area_code") == area_code)
+            result_df = random_distribution(
+                area_df, num_iterations=config.get("num_iterations", 10)
+            )
+            result_dfs.append(result_df)
+
+        final_result_df = pl.concat(result_dfs)
+
+        # Check if visualization is enabled in the config
+        if config.get("enable_visualization", False):
+            visualize_results(final_result_df, area_codes)
+            logger.info("Visualization completed.")
+        else:
+            logger.info("Visualization skipped as per configuration.")
+
+        logger.info("Random distribution completed.")
+        logger.info(final_result_df.head(20))
+    else:
+        logger.info("Random distribution skipped as per configuration.")
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Don't have this here, move it to `scripts/join_sa1_with.....py` and add arguments with `argparse` (see #58 for example)

Reply:
ok this will take a while haha